_posts/2022-01-21-safe-releases.md

```diff
@@ -0,0 +1,95 @@
+---
+layout: post
+title: Reducing thrash and increasing throughput by focusing on elegant releases
+author_avatar: https://avatars.githubusercontent.com/u/8116516?v=4
+author_name: Xavier Hocquet
+author_github_username: xhocquet
+---
+
+Continuous integration and deployment is fundamental to a high-performing technology business as set forth by Kim et al. Deploying frequently results in smaller releases, more frequent updates for your users, quicker feedback from real users, and a tighter feedback loop between business decisions and shipped products.
+
+The worst enemy of continuous deployment is a broken build. The larger the engineering team, the worse it is. If a bad data migration goes out, all releases may have to be paused until a resolution is shipped. Paused releases mean that code branches pile up and inevitably time will be spent fixing issues or rebasing pull requests.
+
+While there are many tools at our disposal to solve some of these problems, I believe a solid understanding of the basics helps us reason about releases and avoid the biggest problems in the first place.
+
+Below I will outline the release cycle for many modern apps (with a focus on Heroku/AWS and Ruby/Javascript), as well as some guidelines to plan graceful releases and migrations.
+
+![Typical release cycle](/images/posts/2022-01-21-safe-releases/release-process.png)
+
+## Release Cycle Steps 1 & 2
+
+Many modern apps follow this 4-step deploy cycle (of course, your setup may vary).
+
+Steps one and two are pretty safe for most apps. You click a deploy button or execute a CLI command (perhaps from your CI tool) which kicks off the build process for your new release. In our apps, that executes a `rails assets:precompile` run which ends up compiling our Javascript assets with webpacker.
+
+## Danger Zone 1 - Executing Release Commands
+
+These are oftentimes data setup steps that need to be in place before new code is running. Things like:
+Database migrations
+Uploading build assets to external sources
+Preparing new environment configurations
+
+This is a danger zone because your data changes need to be backwards-compatible with both versions of code which may be running against your environment. Before the next step, your old code is still being run. After the deployment, your new code will be active.
+
+## Danger Zone 2 - Code Deploy Phase
+
+Finally, your new code is pushed to servers. This can happen in one go or it can be controlled with architectures like blue-green deployments. This is also a danger zone because the new code will fail if your release commands did not prepare the environment properly.
+
```

Comment:
Since you called out that we use Heroku, a "Danger Zone 3" could be the roll Heroku's Preboot will play if you have that enabled. https://devcenter.heroku.com/articles/preboot

This type of thing specifically:
"Whoever is doing the deployment will have to wait a few minutes before the new code starts serving user requests. This happens later than it would without preboot (but in the meantime, user requests are still served by old dynos)."

I don't think it's accurate, but I think preboot works similar to blue-green deploys. But it's more "unpredictable". So in general , safe releases are your best best if you want reliable deploys and behavoir in your apps running on Heroku.